### PR TITLE
kafka: produce request for mesh-filter

### PR DESCRIPTION
**Pull request by adamkotwasinski**
Commit Message: kafka: produce request for mesh-filter
Additional Description: add produce handler for mesh filter; produce handler needs to extract records out of produce requests' bytes, send the records upstream (using Kafka producer), and will wait until it gets notifications (`onDelivery`) to finally find out that the processing has finished, and we can notify the filter (== send the response to downstream Kafka client).
Risk Level: Low
Testing: unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: contrib/kafka/filters/network/source/mesh/command_handlers/produce.cc [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    notifyFilter();
+  }
+}
+
+bool ProduceRequestHolder::finished() const { return 0 == expected_responses_; }
+
+// Find a record that matches provided delivery confirmation coming from Kafka producer.
+// If all the records got their delivery data filled in, we are done, and can notify the origin
+// filter.
+bool ProduceRequestHolder::accept(const DeliveryMemento& memento) {
```
Comment (adamkotwasinski) [inline contrib/kafka/filters/network/source/mesh/command_handlers/produce.cc:50]: [future]
this method gets invoked thru future [RichKafkaProducer implementation](https://github.com/envoyproxy/envoy/pull/11936/files#diff-cf7872d08318648e46c048d2bb615fb4bfc6929386c2f2c4367be65d5936939aR122) going thru the dispatcher to handle [librdkafka delivery confirmation](https://github.com/envoyproxy/envoy/pull/11936/files#diff-cf7872d08318648e46c048d2bb615fb4bfc6929386c2f2c4367be65d5936939aR110)

TLDR: librdkafka callback (dedicated thread) ~> dispatcher ~> RichKafkaProducer (envoy worker thread) -> PRH::accept

**Comment by adamkotwasinski** [comment]
remove label

File: contrib/kafka/filters/network/source/mesh/upstream_kafka_client.h [hunk 2]
```diff
@@ -17,6 +17,13 @@ namespace Mesh {
 // in case of success this means offset (if acks > 0), or error code.
 struct DeliveryMemento {
 
+  // Pointer to byte array that was passed to Kafka producer.
+  // We use this to tell apart messages.
+  // Important: we do not free this memory, it's still part of the 'ProduceRequestHandler' object.
+  // Future work: adopt Kafka's opaque-pointer functionality so we use less memory instead of
+  // keeping whole payload until we receive a confirmation.
```
Comment (adamkotwasinski) [inline contrib/kafka/filters/network/source/mesh/upstream_kafka_client.h:24]: [future work] we'd be sending [with this](https://github.com/edenhill/librdkafka/blob/v1.7.0/src-cpp/rdkafkacpp.h#L3137) and getting back the same pointer in [callback's argument](https://github.com/edenhill/librdkafka/blob/v1.7.0/src-cpp/rdkafkacpp.h#L2309)

**Review by mattklein123 — APPROVED** [review]
